consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/pki/PkiQbftExtraDataCodec.java

```diff
@@ -20,71 +23,67 @@
- We have an "envelope" list, with two elements: the extra data and the cms message.
- The RLP encoding format is as follows: ["extra_data", ["cms"]]
+ The PkiQbftExtraData encoding format is different from the "regular" QbftExtraData encoding. We
+ have an extra bytes element in the end of the list.
 */
 public class PkiQbftExtraDataCodec extends QbftExtraDataCodec {
 
+  public static final int QBFT_EXTRA_DATA_LIST_SIZE = 5;
+
   @Override
   public BftExtraData decodeRaw(final Bytes input) {
     if (input.isEmpty()) {
       throw new IllegalArgumentException("Invalid Bytes supplied - Bft Extra Data required.");
     }
 
-    final RLPInput rlpInput = new BytesValueRLPInput(input, false);
-    rlpInput.enterList();
+    final BftExtraData bftExtraData = super.decodeRaw(input);
 
-    // Consume all the ExtraData input from the envelope list, and decode it using the QBFT decoder
-    final Bytes extraDataListAsBytes = rlpInput.currentListAsBytes();
-    final BftExtraData bftExtraData = super.decodeRaw(extraDataListAsBytes);
+    final RLPInput rlpInput = new BytesValueRLPInput(input, false);
 
-    final Optional<Bytes> cms;
-    if (rlpInput.nextIsList() && rlpInput.nextSize() == 0) {
-      cms = Optional.empty();
+    final Bytes cms;
+    final List<RLPInput> elements = rlpInput.readList(RLPInput::readAsRlp);
+    if (elements.size() > QBFT_EXTRA_DATA_LIST_SIZE) {
+      final RLPInput cmsElement = elements.get(elements.size() - 1);
+      cms = cmsElement.readBytes();
     } else {
-      rlpInput.enterList();
-      cms = Optional.of(rlpInput.readBytes());
-      rlpInput.leaveList();
+      cms = Bytes.EMPTY;
     }
 
     return new PkiQbftExtraData(bftExtraData, cms);
   }
 
   @Override
   protected Bytes encode(final BftExtraData bftExtraData, final EncodingType encodingType) {
+    return encode(bftExtraData, encodingType, true);
+  }
+
+  private Bytes encode(
+      final BftExtraData bftExtraData, final EncodingType encodingType, final boolean includeCms) {
+    final Bytes encoded = super.encode(bftExtraData, encodingType);
     if (!(bftExtraData instanceof PkiQbftExtraData)) {
-      throw new IllegalStateException(
-          "PkiQbftExtraDataCodec must be used only with PkiQbftExtraData");
+      return encoded;
     }
-    final PkiQbftExtraData extraData = (PkiQbftExtraData) bftExtraData;
 
-    final BytesValueRLPOutput encoder = new BytesValueRLPOutput();
-    encoder.startList(); // start envelope list
+    final BytesValueRLPOutput rlpOutput = new BytesValueRLPOutput();
+    rlpOutput.startList();
+    // Read through extraData RLP list elements and write them to the new RLP output
+    new BytesValueRLPInput(encoded, false)
+        .readList(RLPInput::readAsRlp).stream()
+            .map(RLPInput::raw)
+            .forEach(rlpOutput::writeRLPBytes);
+    // Only write CMS if includeCms is true
+    rlpOutput.writeBytes(includeCms ? ((PkiQbftExtraData) bftExtraData).getCms() : Bytes.EMPTY);
```

Comment:
This looks like when encoding without a CMS will add a list item with empty bytes but the decoding allows for the empty CMS to be represented by empty bytes. For consistency be nice if the empty CMS case was represented not being in the list I think.